lib/rugged_adapter/git_layer_rugged.rb

```diff
@@ -165,22 +165,25 @@ def exist?
         ::File.exists?(@repo.path)
       end
 
-      def grep(query, options={})
-        ref = options[:ref] ? options[:ref] : "HEAD"
-        tree = @repo.lookup(sha_from_ref(ref)).tree
-        tree = @repo.lookup(tree[options[:path]][:oid]) if options[:path]
-        enc = options.fetch(:encoding, 'utf-8')
+      def grep(search_terms, options={})
+        ref   = options[:ref] ? options[:ref] : "HEAD"
+        tree  = @repo.lookup(sha_from_ref(ref)).tree
+        tree  = @repo.lookup(tree[options[:path]][:oid]) if options[:path]
+        enc   = options.fetch(:encoding, 'utf-8')
+        query = /^(.*(?:#{search_terms.join('|')}).*)$/i
         results = []
         tree.walk_blobs(:postorder) do |root, entry|
-          blob = @repo.lookup(entry[:oid])
+          blob  = @repo.lookup(entry[:oid])
           count = 0
-          next if blob.binary?
-          blob.content.force_encoding(enc).each_line do |line|
-            next unless line.match(/#{query}/i)
-            count += 1
+          count += 1 if entry[:name] =~ query
+          next if count == 0 && blob.binary?
+          found_in_blob = []
+          blob.content.force_encoding(enc).scan(query) do |match|
```

Comment:
You mean `scan.each`? That would have to be `scan.flatten.eac`. I was hoping this might be a bit faster, as it wouldn't require `scan` to construct the array, which we would then have to flatten. But admittedly that's just guessing about performance

Reply:
`#flatten` or `#compact`? Either way, I don't really know what would be faster, just that `scan.each` seems more idiomatic to me.

Reply:
Agreed, but `.scan do |match|` is at least not unheard of, given`.gsub do |match|` . I'll sleep on it :)